### PR TITLE
Add iterator to read BGZ files line by line

### PR DESCRIPTION
**Pull request by jaudoux**
Hi @brentp ,

Just add an iterator in BGZF lib to easily loop over gzip files line by line. Not sure if this is the best implementation, let me know if there is better way to do it.

Best,
J.

### CODE REVIEW

**Comment by brentp** [comment]
Looks good, except that getline can return -2 for error and that will get swallowed here.
Can you either document that or test the actual return value? -1 is EOF, -2 is error.
On -2, you could raise an IOError

File: src/hts/bgzf.nim [hunk 1]
```diff
@@ -55,13 +55,19 @@ iterator items*(b: BGZ): string =
...
   var p = kstr.addr
-  
-  while b.read_line(p) >= 0:
+
+  r = b.read_line(p)
+  while r >= 0:
     yield $kstr.s
+    r = b.read_line(p)
+  
+  if r <= -2:
```
Comment (brentp) [inline src/hts/bgzf.nim]: this is a nitpick, but could you free before this so that catching the exception won't result in a leak?

**Comment by jaudoux** [comment]
You're right, I did not thought about that. The second commit includes catching of this error as well as a unit test to verify that.

FYI, I first tried to test the error handling with a corrupted gzip file and no error was detected, `bgzf_getline` was returning -1 as a EOF. However the error shows up when the file is compressed with BGZF...

**Comment by brentp** [comment]
yeah, I think we'd have to use `is_bgzf` at opening or whatever the API function is called to detect a gzip (not bgzip) file.

File: src/hts/bgzf.nim (hunk 1 above)
Comment (jaudoux) [reply src/hts/bgzf.nim]: Done. A leak, is a leak 🤓

**Comment by jaudoux** [comment]
yes, but it's usefull to rely on hts-lib for gzip file as well, instead of adding another dependency.

File: tests/bgzftest.nim [hunk 2]
```diff
@@ -31,6 +31,29 @@ suite "bgzf-suite":
...
+      if nb_lines == 2:
+        check line == "b\t10\t20"
+    
+    check nb_lines == 2
+
+  test "bgzf-iterator-err":
+    var 
+      b: BGZ
+      nb_lines = 0
+    b.open("tests/gzip-err.gz", "r")
```
Comment (brentp) [inline tests/bgzftest.nim]: does this need to be added to the repo?

**Comment by brentp** [comment]
> yes, but it's usefull to rely on hts-lib for gzip files as well, instead of adding another dependency.

that could be handled by hts_open, but this is better.

File: tests/bgzftest.nim (hunk 2 above)
Comment (jaudoux) [reply tests/bgzftest.nim]: I forgot to commit the file, done.

**Comment by brentp** [comment]
thanks!

**Pull request merged**